acmed/config/nginx_hooks.toml

```diff
@@ -0,0 +1,130 @@
+  allow all;
+  root "{{#if env.HTTP_ROOT}}{{env.HTTP_ROOT}}{{else}}/var/lib/acmed/domains{{/if}}";
+  default_type "text/plain";
+  try_files $uri =404;
+}
+"""
+stdout = "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_CHALLENGE_LOCATION}}{{env.NGINX_CHALLENGE_LOCATION}}{{else}}001-challenge-letsencrypt.conf{{/if}}"
+
+[[hook]]
+name = "nginx-config-challenge-location-echo"
+type = ["challenge-http-01", "post-operation"]
+cmd = "echo"
+args = [
+    "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_CHALLENGE_LOCATION}}{{env.NGINX_CHALLENGE_LOCATION}}{{else}}001-challenge-letsencrypt.conf{{/if}}"
+]
+allow_failure = true
+
+[[group]]
+name = "nginx-config-challenge-location"
+# hook execution in order of definition
+hooks = [
+    "nginx-config-challenge-location-create",
+    "nginx-config-challenge-location-echo",
+    "nginx-config-challenge-location-chmod"
+]
+
+###
+# nginx reference to TLS certificates
+# env NGINX_TLS_CERTIFICATE -> 002-tls-certificates.conf
+# global: {{certificates_directory}}
+# certificate: {{name}}_{{key_type}}.{{file_type}}.{{ext}}
+###
+
+[[hook]]
+name = "nginx-config-certificate-location-chmod"
+type = ["challenge-http-01", "post-operation"]
+cmd = "chmod"
+args = [
+    "ug+rw",
+    "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_TLS_CERTIFICATE}}{{env.NGINX_TLS_CERTIFICATE}}{{else}}002-tls-certificates.conf{{/if}}"
+]
+allow_failure = true
+
+[[hook]]
+name = "nginx-config-certificate-location-create"
+type = ["challenge-http-01", "post-operation"]
+cmd = "cat"
+args = [ "-" ]
+## TODO
+# make following handlebars accessible in post-operation
+# ssl_certificate {{certificates_directory}}/{{name}}_{{key_type}}.crt.{{ext}};
+# ssl_certificate_key {{certificates_directory}}/{{name}}_{{key_type}}.pk.{{ext}};
+# workaround: define a env:ACMED_CERTS pointing to {{certificates_directory}}
+##
+stdin_str = """###
+# Let's Encrypt TLS certificats
+###
+
+ssl_certificate {{#if env.ACMED_CERTS}}{{env.ACMED_CERTS}}{{else}}/var/lib/acmed/certs{{/if}}/{{identifiers.[0]}}_{{key_type}}.crt.pem; # managed by ACMEd
+ssl_certificate_key {{#if env.ACMED_CERTS}}{{env.ACMED_CERTS}}{{else}}/var/lib/acmed/certs{{/if}}/{{identifiers.[0]}}_{{key_type}}.pk.pem; # managed by ACMEd
+
+ssl_session_cache shared:le_nginx_SSL:10m;
+ssl_session_timeout 1440m;
+ssl_session_tickets off;
+
+ssl_protocols TLSv1.2 TLSv1.3;
+ssl_prefer_server_ciphers off;
+
+ssl_ciphers 'ECDHE-ECDSA-AES128-GCM-SHA256:ECDHE-RSA-AES128-GCM-SHA256:ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:ECDHE-ECDSA-CHACHA20-POLY1305:ECDHE-RSA-CHACHA20-PO
+LY1305:DHE-RSA-AES128-GCM-SHA256:DHE-RSA-AES256-GCM-SHA384';
+"""
+stdout = "{{#if env.NGINX_CONFDIR}}{{env.NGINX_CONFDIR}}{{else}}/etc/nginx/conf.d{{/if}}/{{#if env.NGINX_TLS_CERTIFICATE}}{{env.NGINX_TLS_CERTIFICATE}}{{else}}002-tls-certificates.conf{{/if}}"
```

Comment:
Since the content depends on the domain name, it may be better to fail if the `NGINX_TLS_CERTIFICATE` environment variable isn't defined rather that risking to override an other domain's configuration.

Reply:
Here we go with the chicken-egg problem. 
The intention of the hook tries to provide a valid certificate. The webserver needs to include it to take advantage of the cert. If an admin has choosen to make use of acmed, he should be aware that acmed will request and update the defined certificate if she starts the daemon. If the website config has already been adapting another cert-inclusion this is only solvable, if acmed will parse the website config. And this is out of scope, since it would require acmed to have write access to all config files.
Right now, the hook tries to provide a conf template, that `only` referes to the needed certificate. And this is one file. Your are right, if `NGINX_TLS_CERTIFICATE` is already existing in the filesystem, that would overwrite the given config. But in that case, it is fair to assume that the webserver config is already using/including the conf template. And acmed will provide the updated certificate. No?
And consider, that we are asuming that `acmed` is running on the same host as the webserver. 

If `acmed` will handle multiple domains, the distribution problem will be more complicate. How is `acmed` aware,

* which host will run the webserver for a given domain
* which protocol can be used to deliver the template file (`cp`, `ssh`, `scp` ....)
* which access rights are needed to deliver the template file

This questions are definetly out of scope without a sophisticated parser and automation routines inside `acmed`. I haven't made up my mind if it is worth the effort.
We might deliver the template to `conf.d` subdirectory of the  `domain`. We already handle the write access for the `acmed`.  In conclusion, the default location for `NGINX_CONFDIR` could be changed to `HTTP_ROOT`/domain/`. As an alternative we could introduce an `ACMED_TEMPLATE_DIR`  ...